### PR TITLE
Add CLI Foundry Template (cont.)

### LINKED ISSUE #185: Create CLI Foundry template
## Description

There is a CLI Hardhat template called [`cli-template-contracts-hardhat`](https://github.com/semaphore-protocol/semaphore/tree/main/packages/cli-template-contracts-hardhat) with a sample contract and tests for that contract. 

The idea is to recreate that same template using Foundry instead of Hardhat. The template will be called `cli-template-contracts-foundry`.

## Previous work

https://github.com/vplasencia/semaphore-foundry

https://github.com/vplasencia/semaphore-foundry-hardhat

https://github.com/vplasencia/semaphore-hardhat-foundry

### PR DESCRIPTION
**Pull request by jimmychu0807**
## Description

Completing #854

1. Preferring using `node_modules` instead of git submodules in this repo.
2. As I am making this a js package with package.json, I removed the `Makefile`.

This is a draft PR, but please review the updated approach and see if it is okay. 

A few tasks left:

- [x] Enhance the test so it test for event with exact parameters and verify the contract state.
- [x] Allow running anvil node locally and deploy Feedback.sol to it.
- [x] Update readme.

## Related Issue(s)

closes #185 

## Other information

cc @cedoor @vplasencia @sripwoud 
cc @timou0911 @csiejimmyliu

I thought I could continue building on top of #854, but end up opening a new PR. Let me know if there is a way / how to do it if this is important enough.

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] My changes generate no new warnings
-   [x] I have run `yarn format` and `yarn lint` without getting any errors
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by jimmychu0807** [comment]
I think this PR is ready for review. Thanks.

cc @cedoor @vplasencia @sripwoud
cc @timou0911 @csiejimmyliu

File: packages/cli-template-contracts-foundry/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    cp .env.example .env
+    ```
+
+3. Open the file `apps/web-app/.env.development`. Modify the values of `NEXT_PUBLIC_FEEDBACK_CONTRACT_ADDRESS` and `NEXT_PUBLIC_SEMAPHORE_CONTRACT_ADDRESS` with **feedbackAddr** and **semaphoreAddr** values shown in step 1.
+
+4. Run the Boilerplate front end:
+
+    ```sh
+    yarn dev:web-app
+    ```
```
Comment (jimmychu0807) [inline packages/cli-template-contracts-foundry/README.md:113]: Added a readme section here on integrating this foundry template with Boilerplate front end.

**Comment by jimmychu0807** [comment]
Would like to followup and resolve on this PR. 

@cedoor @sripwoud will you have a chance to take a look at this PR in this week or next? Thanks.

**Comment by vplasencia** [comment]
Hey @jimmychu0807! I will review the PR this week.

**Comment by vplasencia** [comment]
Thank you very much for creating this PR and following up @jimmychu0807 🙏

**Review by vplasencia — CHANGES_REQUESTED** [review]
Thank you so much for this PR @jimmychu0807 🙏

Since the Semaphore templates are configured to use `yarn`, could you add a [`.yarnrc.yml`](https://github.com/semaphore-protocol/semaphore/blob/main/packages/cli-template-contracts-hardhat/.yarnrc.yml) file and add the lines related to npm and yarn to the `.gitignore` file: https://github.com/semaphore-protocol/semaphore/blob/main/packages/cli-template-contracts-hardhat/.gitignore

It would also be nice to have a yarn release in the template like in the others: https://github.com/semaphore-protocol/semaphore/tree/main/packages/cli-template-contracts-hardhat/.yarn/releases

Not all the tests are passing successfully for me, are all the tests passing successfully for you?

File: packages/cli-template-contracts-foundry/remappings.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+@semaphore/contracts/=../../node_modules/@semaphore-protocol/contracts/
+@zk-kit/lean-imt.sol/=../../node_modules/@zk-kit/lean-imt.sol/
+forge-std/=../../node_modules/forge-std/src/
+poseidon-solidity/=../../node_modules/poseidon-solidity/
```
Comment (vplasencia) [inline packages/cli-template-contracts-foundry/remappings.txt]:
```suggestion
@semaphore/contracts/=./node_modules/@semaphore-protocol/contracts/
@zk-kit/lean-imt.sol/=./node_modules/@zk-kit/lean-imt.sol/
forge-std/=./node_modules/forge-std/src/
poseidon-solidity/=./node_modules/poseidon-solidity/
```

Are those remappings working for you? These are the remappings working for me.

**Comment by jimmychu0807** [comment]
@vplasencia, thanks for the review.

> Not all the tests are passing successfully for me, are all the tests passing successfully for you?

hmm... I see. I wonder how do you run the test? The following is my screenshot:

Screenshot 2024-12-16 at 11 17 31

---
p.s. I have updated the gitignore and yarnrc file.

p.p.s. I believe in this monorepo setting, the `node_modules` dir is going to be stored in the root level, not in the package level. That's why the paths in the `remappings.txt` all start with `../..`. I wonder do you take the template-foundry package out to become an independent node project during your test?

**Comment by vplasencia** [comment]
Hey @jimmychu0807!

> hmm... I see. I wonder how do you run the test? The following is my screenshot:

One of my tests was failing because I had to run `foundryup` to update the Foundry version. Now they work fine after modifying the remappings.

> p.p.s. I believe in this monorepo setting, the node_modules dir is going to be stored in the root level, not in the package level. That's why the paths in the remappings.txt all start with ../... I wonder do you take the template-foundry package out to become an independent node project during your test?

Yes, the templates should be tested outside because they will be independent projects. This will be executed to create a new project using a template: https://github.com/semaphore-protocol/semaphore/blob/main/packages/cli/src/index.ts#L54

**Review by vplasencia — COMMENTED** [review]
It would be nice to add the Foundry template to the CLI so that developers can use the template right after it's ready. You can do this by adding:

```ts
    {
        value: "contracts-foundry",
        name: "Foundry"
    }
```

to the `supportedTemplates` here: https://github.com/semaphore-protocol/semaphore/blob/main/packages/cli/src/index.ts#L23-L36

**Comment by jimmychu0807** [comment]
> Yes, the templates should be tested outside because they will be independent projects. This will be executed to create a new project using a template: https://github.com/semaphore-protocol/semaphore/blob/main/packages/cli/src/index.ts#L54

ah, I see the context now. Let me update it accordingly.

**Comment by jimmychu0807** [comment]
@vplasencia Updated the packages/cli and remapping path.

---
I wonder how do you test the packages/cli locally? I run the command `yarn workspace @semaphore-protocol/cli start create` and create a project using `contracts-foundry (Foundry)` template. At some point, it will try to download the package from https://registry.npmjs.org and complain it is not found (as it is not published yet). How do you make it using the local foundry template instead of from npm?

File: packages/cli-template-contracts-foundry/remappings.txt (hunk 2 above)
Comment (jimmychu0807) [reply packages/cli-template-contracts-foundry/remappings.txt]: updated

**Comment by vplasencia** [comment]
Hey @jimmychu0807 thank you very much for all the updates.

> I wonder how do you test the packages/cli locally? I run the command yarn workspace @semaphore-protocol/cli start create and create a project using contracts-foundry (Foundry) template. At some point, it will try to download the package from https://registry.npmjs.org/ and complain it is not found (as it is not published yet). How do you make it using the local foundry template instead of from npm?

To test the new template you can just copy/paste the `cli-template-contracts-foundry` folder somewhere else and add the `yarn.lock` file.

**Review by vplasencia — APPROVED** [review]
Thank you very much for the great work @jimmychu0807 🚀

**Pull request merged**